uber/jinja.py

```diff
@@ -0,0 +1,61 @@
+from uber.common import *
+
+
+class JinjaEnv:
+    _env = None
+    _exportable_functions = {}
+    _filter_functions = {}
+    _template_dirs = []
+
+    @staticmethod
+    def append_template_dir(dirname):
+        """
+        Add another template directory we should search when looking up templates.
+        """
+        JinjaEnv._template_dirs.insert(0, dirname)
+
+    @staticmethod
+    def env():
+        if JinjaEnv._env is None:
+            JinjaEnv._env = JinjaEnv._init_env()
+        return JinjaEnv._env
+
+    @staticmethod
+    def _init_env():
+        env = jinja2.Environment(
+                # autoescape=_guess_autoescape,
+                loader=jinja2.FileSystemLoader(JinjaEnv._template_dirs)
+            )
+
+        for name, func in JinjaEnv._exportable_functions.items():
+            env.globals[name] = func
+
+        for name, func in JinjaEnv._filter_functions.items():
+            env.filters[name] = func
+
+        return env
+
+    @staticmethod
+    def jinja_export(name=None):
+        def wrap(func):
+            JinjaEnv._exportable_functions[name if name else func.__name__] = func
+            return func
+        return wrap
+
+    @staticmethod
+    def jinja_filter(name=None):
```

Comment:
Incidentally, what we're doing here is a somewhat common pattern, which is "I want a decorator that sometimes takes an argument but sometimes doesn't."

There are two ways to go about this.  The default way is to do what you did here, i.e. force people to always say ``@JinjaEnv.jinja_filter()`` with the parenthesis at the end.

The other way is to check whether or not the thing we've been passed is a function and do things differently based on whether or not it is, i.e.

```python
@classmethod
def jinja_filter(arg=None):
    def _register(func, name=None):
        cls._filter_functions[name if name else func.__name__] = func

    if isinstance(arg, FunctionType):  # have to do "from types import FunctionType"
        _register(arg)
        return arg
    else:
        def registrar(func):
            _register(func, arg)
            return func
        return registrar
```

The above code allows us to say either ``@JinjaEnv.jinja_filter`` or ``@JinjaEnv.jinja_filter('filter_name')`` without any issue.  It even supports people putting the parenthesis on the end when they don't have to, i.e. ``@JinjaEnv.jinja_filter()`` like we're doing now.